File: 3sum.cc
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> solns;

        for (int i = 0; i < nums.size() - 2; i++) {
            if (nums[i] > 0) break;
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            int l = i + 1;
            int r = nums.size() - 1;
            while (l < r) {
                int sum = nums[l] + nums[r] +nums[i];
                if (sum > 0) r--;
                else if (sum < 0) l++;
                else {
                    solns.push_back({nums[i], nums[l], nums[r]});
                    while (l < r && nums[l] == nums[l + 1]) l++;
                    while (l < r && nums[r] == nums[r - 1]) r--;
                    l++; r--;
                }
            }
        }

        return solns;
    }
};

File: python/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        sortedNums = sorted(nums)
        unique_combos = []

        for i in range(len(sortedNums) - 2):
            if i > 0 and sortedNums[i] == sortedNums[i - 1]:
                continue
            left = i + 1
            right = len(nums) - 1
            while left < right:
                total = sortedNums[i] + sortedNums[left] + sortedNums[right]
                if total == 0:
                    unique_combos.append([sortedNums[i], sortedNums[left], sortedNums[right]])
                    left += 1
                    right -= 1

                    while left < right and sortedNums[left] == sortedNums[left - 1]:
                        left += 1
                    while left < right and sortedNums[right] == sortedNums[right + 1]:
                        right -= 1
                elif total < 0:
                    left += 1
                elif total > 0:
                    right -= 1
        return unique_combos
